File: ruby/grains/grains.rb
class Grains
  def self.square(num)
    raise ArgumentError if num <= 0 || num > 64
    self.chess_board_grains(num).last
  end

  def self.total
    self.chess_board_grains.inject(0, :+)
  end

  def self.chess_board_grains(num=64)
    (2..num).each_with_object(Array.new(1, 1)) do |i, array|
      array << 2 * array.last
    end
  end
end

File: go/grains/grains.go
package grains

import "fmt"

func Square(square int) (uint64, error) {
	if square < 1 || square > 64 {
		return 0, fmt.Errorf("Not a valid space: %v", square)
	}

	var total uint64 = 1
	for i := 2; i <= square; i++ {
		total *= 2
	}

	return total, nil
}

func Total() uint64 {
	var total uint64
	for i := 1; i <= 64; i++ {
		val, _ := Square(i)
		total += val
	}
	return total
}
